spinalcordtoolbox/image.py

```diff
@@ -1704,3 +1707,66 @@ def compute_cross_corr_3d(image: Image, coord, xrange=list(range(-10, 10)), xshi
     # Change adjust rl_coord
     logger.info('R-L coordinate adjusted from %s to  %s)', x, x + xrange[ind_peak])
     return x + xrange[ind_peak]
+
+
+def stitch_images(fnames_in: list, fname_out='stitched.nii.gz'):
+    """
+    Stitch two (or more) images utilizing the C++-precompiled binaries of Biomedia-MIRA's stitching toolkit
+    (https://github.com/biomedia-mira/stitching) by placing a system call.
+
+    :param fnames_in: list of filenames of scans in RPI orientation. e.g. [stack1.nii.gz, stack2.nii.gz]
+    :param fname_out: filename for stitched output scan.
+    :param directory: path to temp directory where we store the result of stitching.nii.gz
+    :return: none
+    """
+    # stringify the fname list to parsable cmd parameter
+    fnames_cmd = " ".join(fnames_in)
+
+    cmd = ['isct_stitching', '-i', fnames_cmd, '-o', fname_out, '-a']
+    status, output = run_proc(cmd, verbose='verbose', is_sct_binary=True)
+    print(status, output)
+    if status != 0:
+        raise RuntimeError(f"Subprocess call {cmd} returned non-zero: {output}")
+
+    return Image(fname_out)
+
+
+def generate_stitched_qc_images(ims_in: Sequence[Image], im_out: Image):
+    """
+    Pad input and output images, so that a QC report can compare between the output of
+    'stitching', vs. the naive output of concatenating the input images together.
+    """
+    # find the max shape of all input images
+    shape_max = [max(im.data.shape[0] for im in ims_in),
+                 max(im.data.shape[1] for im in ims_in),
+                 max(im.data.shape[2] for im in ims_in)]
+
+    # pad any input images that are smaller than the max shape
+    # (the stitching tool can handle mismatched image shapes natively, but we have to manage it ourselves)
+    for im in ims_in:
+        for i in [0, 1]:
+            diff = shape_max[i] - im.data.shape[i]
+            if diff:
+                padding = [[0, 0], [0, 0], [0, 0]]
+                padding[i][0] = round(diff / 2)
+                padding[i][1] = diff // 2
+                im.data = np.pad(im.data, padding)
+
+    # create a 1-voxel blank image for visual clarity
+    im_blank = Image([shape_max[0], shape_max[1], 1])
+
+    # create a naively-stitched (RPI) image for comparison in QC report
+    # NB: we reverse the list of images because numpy's origin location (bottom) is are different than nibabel's (top)
+    im_concat_list = list(reversed(ims_in))
+    for i in range(1, len(im_concat_list), 2):  # insert blank 'spacer' images in between each image
+        im_concat_list.insert(i, im_blank)
```

Comment:
This only inserts half of the necessary spacer images, unfortunately:

```py3
>>> my_list = list(range(10))
>>> for i in range(1, len(my_list), 2): my_list.insert(i, None)
... 
>>> my_list
[0, None, 1, None, 2, None, 3, None, 4, None, 5, 6, 7, 8, 9]
```

I wish Python had something like `str.join`, but for lists, because that's exactly what we want here. It's surprisingly awkward to write. The most straightforward way I could come up with is this:
```suggestion
    im_concat_list = []
    for im in reversed(ims_in):
        # insert blank 'spacer' images in between each image
        im_concat_list.extend([im, im_blank])
    if im_concat_list:
        # remove the trailing 'spacer', if there is one
        im_concat_list.pop()
```

(The `if im_concat_list:` test at the end is needed in case `ims_in` was empty, otherwise `im_concat_list.pop()` would raise `IndexError: pop from empty list`. We probably don't expect `ims_in` to be empty in practice, but I like to cover my bases.)

Reply:
Oh gosh, what an embarrassing mistake on my part! Thank you for acknowledging this. 

That said, I think we can still use `range` here, and avoid the need to remove a trailing spacer? This seems to work for me:

```python
for i in range(1, (len(my_list)*2)-1, 2): 
    my_list.insert(i, None)
```

<details>

```py3
for list_len in range(10):
    print(f"List length: {list_len}")
    my_list = list(range(list_len))
    print(my_list)
    for i in range(1, (len(my_list) * 2) - 1, 2): my_list.insert(i, None)
    print(my_list)
    print()
    
List length: 0
[]
[]
List length: 1
[0]
[0]
List length: 2
[0, 1]
[0, None, 1]
List length: 3
[0, 1, 2]
[0, None, 1, None, 2]
List length: 4
[0, 1, 2, 3]
[0, None, 1, None, 2, None, 3]
List length: 5
[0, 1, 2, 3, 4]
[0, None, 1, None, 2, None, 3, None, 4]
List length: 6
[0, 1, 2, 3, 4, 5]
[0, None, 1, None, 2, None, 3, None, 4, None, 5]
List length: 7
[0, 1, 2, 3, 4, 5, 6]
[0, None, 1, None, 2, None, 3, None, 4, None, 5, None, 6]
List length: 8
[0, 1, 2, 3, 4, 5, 6, 7]
[0, None, 1, None, 2, None, 3, None, 4, None, 5, None, 6, None, 7]
List length: 9
[0, 1, 2, 3, 4, 5, 6, 7, 8]
[0, None, 1, None, 2, None, 3, None, 4, None, 5, None, 6, None, 7, None, 8]
```

</details>

(This is what I get for not thinking carefully enough about [a random SO answer](https://stackoverflow.com/a/66126709).)

Reply:
Hey we all make mistakes, that's what code review is for! That and bugfixes :slightly_smiling_face:

Reply:
> ```python
> for i in range(1, (len(my_list)*2)-1, 2): 
>     my_list.insert(i, None)
> ```

@mguaypaq mentioned this to me in private, but, the solution that I proposed above is accidentally quadratic. 

_((Python lists are optimized for appending, and prepending/insertion require the entire list to be recreated from scratch. And, my proposal inserts a new element on each loop... :sweat_smile:))_

During our conversation, though, @mguaypaq came up with a clever (and quick!) solution: pre-allocating a list of length `2n-1`, then using slicing to assign the images to their correct indexes:

```python
    im_concat_list = [im_blank] * (2*len(ims_in_padded)-1)    # Preallocate a list of blank spacer images
    im_concat_list[::2] = reversed(ims_in_padded)             # Assign the actual images in positions 0, 2, 4, etc.
    im_concat = concat_data(im_concat_list, dim=2)            # Concatenate the images together
```

This is quicker and cleaner than both previous solutions, IMO!

Reply:
Also, I just tested and it looks like this preallocation trick correctly handles the case where `ims_in_padded` is empty. (I was a little surprised that `[im_blank] * (-1)` is `[]` rather than being an error, but it happens to be the right thing here.)